libs/common/ring_buffer.hpp

```diff
@@ -0,0 +1,152 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef IROHA_COMMON_RING_BUFFER_HPP
+#define IROHA_COMMON_RING_BUFFER_HPP
+
+namespace iroha {
+
+    template<typename T, size_t Count> class RingBuffer final {
+    public:
+        using type = T;
+        using handle = size_t;
+
+    private:
+        static_assert(Count > 0, "Unexpected count value. It must be above 0.");
+
+        enum { kSizeType = sizeof(type) };
+        enum { kActualLimit = Count };
+        enum { kCapLimit = kActualLimit + 1 };
+        enum { kVirtualLimit = 2 * kActualLimit };
+
+        struct Node {
+            uint8_t data[kSizeType];
+        } data_[Count];
+
+        handle end_;
+        handle begin_;
+
+        inline bool handleInBound_(handle register h) const {
+            /**
+             * this code is only for debug purpose
+            **/
+            auto register vbegin = begin_ + kVirtualLimit;
+            h += kVirtualLimit;
+            return ((h >= end_) && (h < vbegin));
+        }
+
+        inline size_t incrementAndNormilize(size_t val) const {
+            return (++val % kVirtualLimit);
+        }
+
+        inline size_t handle2Position(handle h) const {
+            return (h % kActualLimit);
+        }
+
+        inline size_t size_() const {
+            assert(((begin_ - end_) % kVirtualLimit) < kCapLimit);
+            return ((begin_ - end_) % kVirtualLimit);
+        }
+
+        inline bool empty_() const {
+            return (begin_ == end_);
+        }
+
+        inline type& getItem_(Node& node) {
+            return *(type*)node.data;
+        }
+
+        inline type const& getItem_(Node const& node) const {
+            return *(type const*)node.data;
+        }
+
+        inline void destruct_(Node& node) {
+            assert(!empty_());
+
+            auto& item = getItem_(node);
+            item.~type();
+            end_ = incrementAndNormilize(end_);
+        }
+
+        template<typename...__args> inline void construct_(Node& node, __args&&... args) {
```

Comment:
this method works under assertion that `node` refers to the first element, so let's place an assertment
```suggestion
        template<typename...__args> inline void constructFirst_(Node& node, __args&&... args) {
          assert(&node == static_cast<Node*>(data_ + handle2Position(begin_)));
```